src/signer/src/main/java/ee/ria/xroad/signer/tokenmanager/token/SoftwareTokenUtil.java

```diff
@@ -86,6 +116,70 @@ public static String getKeyStoreFileName(String keyId) {
         return getKeyDir() + "/" + keyId + P12;
     }
 
+    /**
+     * Returns a temporary filepath for a key store. Used e.g. when changing pin codes for key stores.
+     * Usually the temp dir exists already when this method is called
+     * (to create the temp dir use {@link #createTempKeyDir()})
+     * @param tempKeyDir the temporary key dir path
+     * @param keyId the key id
+     * @return the key store file name for a key id in a temporary folder
+     */
+    public static String getTempKeyStoreFileName(Path tempKeyDir, String keyId) {
+        return tempKeyDir.toString() + File.separator + keyId + P12;
+    }
+
+    /**
+     * Rename the temp dir to the official key dir. Used e.g. when changing pin codes for key stores
+     * @param tempKeyDir the temporary key dir path
+     * @throws IOException renaming fails
+     */
+    public static void renameTempToKeyDir(Path tempKeyDir) throws IOException {
+        Files.move(tempKeyDir, getKeyDir().toPath(), ATOMIC_MOVE);
+    }
+
+    /**
+     * @return /path/to/.softtoken.bak
+     */
+    public static String getBackupKeyDir() {
+        return ResourceUtils.getFullPathFromFileName(KEY_CONF_FILE)
+                + SOFT_TOKEN_KEY_BAK_DIR_NAME + File.separator;
+    }
+
+    /**
+     * @return key backup dir with timestamp in the name e.g. /path/to/.softtoken.bak-20210218102059
+     */
+    public static String getBackupKeyDirForDateNow() {
+        Timestamp nowTimestamp = new Timestamp(System.currentTimeMillis());
+        String nowString = TIMESTAMP_FORMAT.format(nowTimestamp);
+        return ResourceUtils.getFullPathFromFileName(KEY_CONF_FILE)
+                + SOFT_TOKEN_KEY_BAK_DIR_NAME + "-" + nowString + File.separator;
+    }
+
+    /**
+     * Create a temp directory for key stores. Used e.g. when changing pin codes for key stores
+     * @throws IOException creating temp dir fails
+     */
+    public static Path createTempKeyDir() throws IOException {
+        return Files.createTempDirectory(Paths.get(ResourceUtils.getFullPathFromFileName(KEY_CONF_FILE)),
+                SOFT_TOKEN_KEY_DIR_NAME + "-", SOFT_TOKEN_KEY_DIR_PERMISSIONS);
+    }
+
+    /**
+     * Rename the key dir to a backup dir. Used e.g. when changing pin codes for key stores
+     * @throws IOException renaming fails
+     */
+    public static void renameKeyDirToBackup() throws IOException {
+        Files.move(getKeyDir().toPath(), Paths.get(getBackupKeyDir()), ATOMIC_MOVE);
+    }
+
+    /**
+     * Remove the backup key dir. Used e.g. when changing pin codes for key stores
+     * @throws IOException removing fails
+     */
+    public static void removeBackupKeyDir() throws IOException {
+        FileUtils.deleteDirectory(new File(getBackupKeyDir()));
+    }
+
```

Comment:
This utility code seems to use a mixture of Strings, Files, and Paths to represent a (abstract) file or directory. It would make sense to refactor the code to use e.g. Path throughout, because it is a more modern API and composes better. Also some of the utility functions are too specific, e.g. there are two rename "functions" that do the basically the same operation, but some of the parameters are encoded in the function name. Both could be removed and usage replaced with the Files.move() call.

Reply:
Yeah there was some room for streamlining. Most of these methods were functioning just as tiny wrappers so removed most and refactored the remaining to use `Path`.

 I did not touch the original methods (e.g. `getKeyStoreFileName`) because they are used in quite many places and did not want to refactor something that has worked for years.